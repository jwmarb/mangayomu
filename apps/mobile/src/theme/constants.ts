import { moderateScale } from 'react-native-size-matters';

export const BOOK_DIMENSIONS = {
  width: moderateScale(110),
  height: moderateScale(205),
};

export const BOOK_DIMENSIONS_RATIO =
  BOOK_DIMENSIONS.width / BOOK_DIMENSIONS.height;

export const BOOK_COVER_HEIGHT = moderateScale(160);

export const UNFINISHED_MANGA_HEIGHT = moderateScale(128);
export const UNFINISHED_MANGA_WIDTH = moderateScale(267);

export const ACCORDION_SECTION_HEADER_HEIGHT = moderateScale(48);
export const ACCORDION_ITEM_HEIGHT = moderateScale(52);

export const OVERLAY_COLOR = 'rgba(128, 128, 128, 0.5)';
export const OVERLAY_TEXT_PRIMARY = 'rgba(255, 255, 255, 1)';
export const OVERLAY_TEXT_SECONDARY = 'rgba(255, 255, 255, 0.7)';
export const OVERLAY_SLIDER_HEIGHT = moderateScale(45);
export const OVERLAY_HEADER_HEIGHT = moderateScale(64);
export const OVERLAY_SLIDER_CIRCLE_RADIUS = moderateScale(11);
export const OVERLAY_SLIDER_CIRCLE_RIPPLE_RADIUS = moderateScale(16);
export const READER_NETWORK_TOAST_HEIGHT = moderateScale(24);

export const RADIO_BUTTON_BORDER_WIDTH = moderateScale(2);

export const DIVIDER_DEPTH = moderateScale(1.5);

export const LOADING_BAR_HEIGHT = moderateScale(3);
