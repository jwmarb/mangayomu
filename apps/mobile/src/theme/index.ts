export { typography, __storybook_theme__ } from './theme';
