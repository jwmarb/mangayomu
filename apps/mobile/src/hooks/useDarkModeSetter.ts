import React from 'react';
import { ThemeSetDarkModeContext } from '@/providers/theme';

/**
 * A hook that gets the setter for isDarkMode
 * @returns Returns a setter for modifying isDarkMode
 */
export default function useDarkModeSetter() {
  const context = React.useContext(ThemeSetDarkModeContext);
  if (context == null)
    throw new Error(
      'Tried to consume `ThemeDarkModeContext` when component is not a child of it',
    );

  return context;
}
