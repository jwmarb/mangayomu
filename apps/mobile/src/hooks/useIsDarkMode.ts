import React from 'react';
import { ThemeDarkModeContext } from '../providers/theme';

/**
 * A hook that determines if app is in dark mode
 * @returns Returns a boolean indicating whether dark mode is in use or not
 */
export default function useIsDarkMode() {
  const context = React.useContext(ThemeDarkModeContext);
  return context;
}
