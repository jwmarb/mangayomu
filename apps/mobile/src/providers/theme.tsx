import React from 'react';
import { useMMKVBoolean } from 'react-native-mmkv';
import { Theme, createTheme } from '@mangayomu/theme';
import { Appearance, StatusBar, ViewStyle, useColorScheme } from 'react-native';
import {
  NavigationContainer,
  Theme as NavigationTheme,
} from '@react-navigation/native';
import { mmkv } from '@/utils/persist';

declare module '@mangayomu/theme' {
  export interface ThemeHelpers extends DefaultThemeHelpers {
    elevation(n: number): ViewStyle;
  }
  export interface Theme extends Pick<DefaultTheme, 'mode'> {
    palette: DefaultTheme['palette'] & {
      skeleton: string;
      overlay: string;
      backdrop: string;
      divider: string;
      common: {
        white: string;
        black: string;
      };
      action: {
        disabled: string;
        ripple: string;
        textInput: string;
      };
      status: {
        ongoing: string;
        hiatus: string;
        discontinued: string;
        completed: string;
        cancelled: string;
      };
      success: {
        light: string;
        main: string;
        dark: string;
      };
      background: DefaultTheme['palette']['background'] & {
        menu: string;
      };
    };
    helpers: ThemeHelpers;
    style: {
      rippleRadius: number;
      /**
       * @default 32
       * @deprecated use `borderRadius`
       * ```js
       *  borderRadius: {
       *    xxl: 32,
       *     xl: 24,
       *      l: 16,
       *      m: 8,
       *      s: 4
       *  }
       * ```
       */
      borderRadius_legacy: number;
      screen: {
        paddingVertical: number;
        paddingHorizontal: number;
      };
      container: {
        paddingVertical: number;
        paddingHorizontal: number;
      };
      size: {
        s: 4;
        m: 8;
        l: 12;
        xl: 16;
        xxl: 20;
      };
      borderWidth: {
        s: 1.5;
        m: 2.5;
        l: 4;
        xl: 6;
        xxl: 8;
      };
      borderRadius: {
        xxl: 32;
        xl: 24;
        l: 16;
        m: 8;
        s: 4;
      };
    };
    reactNavigation: NavigationTheme;
  }
}

export type ThemeProviderProps = React.PropsWithChildren;

export const ThemeDarkModeContext = React.createContext<boolean>(
  Appearance.getColorScheme() === 'dark',
);
export const ThemeSetDarkModeContext = React.createContext<
  | ((
      value:
        | boolean
        | ((current: boolean | undefined) => boolean | undefined)
        | undefined,
    ) => void)
  | null
>(null);

export const { opposite: lightTheme, ...darkTheme } = createTheme<Theme>(
  ({ color, colorConstant }) => ({
    mode: 'dark' as const,
    helpers: {
      elevation(n: number) {
        return (theme: Theme) =>
          ({
            shadowColor: theme.palette.common.black,
            shadowOffset: {
              width: 0,
              height: n >> 1,
            },
            shadowOpacity: 0.16 + n * 0.2,
            shadowRadius: n / 1.529,
            elevation: n,
          } as ViewStyle);
      },
    },
    palette: {
      overlay: colorConstant('rgba(0, 0, 0, 0.5)'), // experimental
      backdrop: colorConstant('rgba(0, 0, 0, 0.3)'),
      skeleton: color('rgba(255, 255, 255, 0.12)', 'rgba(0, 0, 0, 0.12)'),
      divider: color('rgba(255, 255, 255, 0.12)', 'rgba(0, 0, 0, 0.12)'),
      common: {
        white: colorConstant('#fff'),
        black: colorConstant('#000'),
      },
      action: {
        ripple: color('#606060', '#c2c2c2'),
        disabled: color('rgba(255, 255, 255, 0.12)', 'rgba(0, 0, 0, 0.12)'),
        textInput: color('rgba(128, 128, 128, 0.12)', 'rgba(61, 61, 61, 0.12)'),
      },
      primary: {
        light: color('#B2CBE6', '#34A1FB'),
        main: color('#8DB1D8', '#1996FD'),
        dark: color('#6897CA', '#066EC4'),
      },
      secondary: {
        light: color('#DAA2B7', '#EC8F43'),
        main: color('#D290A9', '#E57417'),
        dark: color('#BA5B7F', '#B55D14'),
      },
      error: {
        main: color('#883934', '#e65252'),
        light: color('#e99090', '#e7a5a4'),
        dark: color('#553c3c', '#3b2020'),
      },
      success: {
        light: color('#d2f3c0', '#E1F7CF'),
        main: color('#71A057', '#b7eb8f'),
        dark: color('#3e4b36', '#72D129'),
      },
      warning: {
        main: color('#ed6c02', '#ffa726'),
        light: color('#ff9800', '#ffb74d'),
        dark: color('#e65100', '#f57c00'),
      },
      text: {
        primary: color('rgba(255, 255, 255, 1)', 'rgba(0, 0, 0, 0.87)'),
        secondary: color('rgba(255, 255, 255, 0.7)', 'rgba(0, 0, 0, 0.6)'),
        disabled: color('rgba(255, 255, 255, 0.3)', 'rgba(0, 0, 0, 0.26)'),
        hint: color('rgba(255, 255, 255, 0.5)', 'rgba(0, 0, 0, 0.38)'),
      },
      background: {
        default: color('#071113', '#fafafa'),
        paper: color('#09181a', '#ffffff'),
        menu: color('#0e2629', '#eeeef1'),
        disabled: color('#102A2D', '#EFEFEF'),
      },
      status: {
        ongoing: color('#73d13d', '#b7eb8f'),
        hiatus: color('#ffc53d', '#ffe58f'),
        discontinued: color('#f5222d', '#ff7875'),
        cancelled: color('#f5222d', '#ff7875'),
        completed: color('#434343', '#8c8c8c'),
      },
    },
    style: {
      rippleRadius: 16,
      borderRadius_legacy: 32,
      borderRadius: {
        xxl: 32,
        xl: 24,
        l: 16,
        m: 8,
        s: 4,
      },
      borderWidth: {
        s: 1.5,
        m: 2.5,
        l: 4,
        xl: 6,
        xxl: 8,
      },
      screen: {
        paddingVertical: (theme) => theme.style.size.s,
        paddingHorizontal: (theme) => theme.style.size.l,
      },
      container: {
        paddingVertical: (theme) => theme.style.size.s,
        paddingHorizontal: (theme) => theme.style.size.m,
      },
      size: {
        s: 4,
        m: 8,
        l: 12,
        xl: 16,
        xxl: 20,
      },
    },
    reactNavigation: {
      dark: (theme) => theme.mode === 'dark',
      colors: {
        background: (theme) => theme.palette.background.default,
        primary: (theme) => theme.palette.primary.main,
        text: (theme) => theme.palette.text.primary,
        border: (theme) => theme.palette.divider,
        card: (theme) => theme.palette.background.paper,
        notification: (theme) => theme.palette.secondary.main,
      },
    },
  }),
);

/**
 * Gets the theme, however, mode-dependent values are static and do not change
 */
export function dangerouslyGetTheme() {
  return Appearance.getColorScheme() === 'dark' ? darkTheme : lightTheme;
}

export function ThemeProvider({ children }: ThemeProviderProps) {
  const [isDarkMode, setIsDarkMode] = useMMKVBoolean('theme', mmkv);
  const systemPreference = useColorScheme();
  const isEffectivelyDarkMode = isDarkMode ?? systemPreference === 'dark';

  React.useEffect(() => {
    if (isEffectivelyDarkMode) StatusBar.setBarStyle('light-content');
    else StatusBar.setBarStyle('dark-content');
  }, [isEffectivelyDarkMode]);

  const navigationTheme = (isEffectivelyDarkMode ? darkTheme : lightTheme)
    .reactNavigation;

  return (
    <NavigationContainer theme={navigationTheme}>
      <ThemeDarkModeContext.Provider value={isEffectivelyDarkMode}>
        <ThemeSetDarkModeContext.Provider value={setIsDarkMode}>
          {children}
        </ThemeSetDarkModeContext.Provider>
      </ThemeDarkModeContext.Provider>
    </NavigationContainer>
  );
}
