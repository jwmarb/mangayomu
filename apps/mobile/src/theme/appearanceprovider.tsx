import React from 'react';
import { ColorSchemeName, StatusBar, useColorScheme } from 'react-native';

import {
  NavigationContainer,
  DefaultTheme as NavigationDefaultTheme,
} from '@react-navigation/native';
import { createTheme, UserDefinedPalette } from '@mangayomu/theme';
import { Theme, ThemeProvider, useTheme } from '@emotion/react';
import { shadow, spacing, typography } from '@theme/theme';
import { helpers } from '@theme/helpers';
import { moderateScale } from 'react-native-size-matters';
import { useMMKV } from 'react-native-mmkv';
import { MenuProvider } from 'react-native-popup-menu';
import DialogProvider from '@components/Dialog/DialogProvider';
import { PortalProvider } from '@gorhom/portal';

export enum AppearanceMode {
  SYSTEM = 'System',
  LIGHT = 'Light',
  DARK = 'Dark',
}

export const useAppearanceColorScheme = (
  mode: AppearanceMode,
): ColorSchemeName => {
  const theme = useColorScheme();
  switch (mode) {
    case AppearanceMode.SYSTEM:
      return theme;
    case AppearanceMode.LIGHT:
      return 'light';
    case AppearanceMode.DARK:
      return 'dark';
  }
};

export const AppearanceContext = React.createContext<
  | {
      mode: AppearanceMode;
      setMode: (setter: AppearanceMode) => void;
    }
  | undefined
>(undefined);

export const useAppearanceMode = () => {
  const ctx = React.useContext(AppearanceContext);
  if (ctx == null)
    throw Error('This component is not a child of AppearanceContext');
  return ctx;
};

export const AppearanceProvider: React.FC<React.PropsWithChildren> = ({
  children,
}) => {
  const theme = useTheme();

  return (
    <MenuProvider>
      <DialogProvider>
        <PortalProvider>
          <NavigationContainer theme={theme.__react_navigation__}>
            {children}
          </NavigationContainer>
        </PortalProvider>
      </DialogProvider>
    </MenuProvider>
  );
};
