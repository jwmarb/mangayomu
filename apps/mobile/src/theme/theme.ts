import { DefaultTheme } from '@mangayomu/theme';
import '@emotion/react';
import { css } from '@emotion/native';
import { RFValue } from 'react-native-responsive-fontsize';

export type TextVariants = keyof typeof typography;

export const typography = {
  header: css`
    font-family: Roboto;
    font-size: ${RFValue(20)}px;
    letter-spacing: -0.2px;
  `,
  body: css`
    font-family: Roboto;
    font-size: ${RFValue(13)}px;
    letter-spacing: -0.44px;
  `,
  button: css`
    font-family: Roboto;
    font-size: ${RFValue(12)}px;
    letter-spacing: -0.44px;
  `,
  'bottom-tab': css`
    font-family: Roboto;
    font-size: ${RFValue(10)}px;
    letter-spacing: -0.56px;
  `,
} as const;

declare module '@emotion/react' {
  export interface Theme extends DefaultTheme {
    typography: typeof typography;
  }
}
