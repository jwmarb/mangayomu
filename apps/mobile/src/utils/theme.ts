/* eslint-disable @typescript-eslint/no-explicit-any */
import { StyleSheet } from 'react-native';
import { Theme } from '@mangayomu/theme';
import { lightTheme, darkTheme } from '@/providers/theme';

/**
 * Same as `StyleSheet.create` except the theme can be passed into it to create theme-dependent styles
 * @param styles A callback function with the theme passed as an argument that returns styles
 * @returns Returns an array with a light and dark theme styles
 */
export function createStyles<
  T extends StyleSheet.NamedStyles<any> | StyleSheet.NamedStyles<T>,
>(styles: (theme: Theme) => T) {
  return [
    StyleSheet.create(styles(lightTheme)),
    StyleSheet.create(styles(darkTheme)),
  ] as const;
}

/**
 * Creates props with a theme imbued with it
 * @param withProps A callback function with the theme passed as an argument that returns props
 * @returns Returns an array with a light theme and a dark theme props.
 */
export function createThemedProps<T>(withProps: (theme: Theme) => T) {
  return [withProps(lightTheme), withProps(darkTheme)] as const;
}
