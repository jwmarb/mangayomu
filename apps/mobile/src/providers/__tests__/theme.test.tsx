import 'react-native';
import { fireEvent, render, screen } from '@testing-library/react-native';
import { ThemeProvider } from '../theme';
import useIsDarkMode from '../../hooks/useIsDarkMode';
import { Button, Text, View } from 'react-native';
import useDarkModeSetter from '../../hooks/useDarkModeSetter';
import React from 'react';
import { Appearance } from 'react-native';

jest.mock('react-native-mmkv', () => ({
  useMMKVBoolean: () => {
    return jest.requireActual('react').useState();
  },
  MMKV: class MMKV {},
}));

function ConsumerComponent() {
  const isDarkMode = useIsDarkMode();
  const setIsDarkMode = useDarkModeSetter();

  return (
    <View>
      <Text testID="isDarkMode">{isDarkMode ? 'Dark Mode' : 'Light Mode'}</Text>
      <Button title="toggle" onPress={() => setIsDarkMode((prev) => !prev)} />
    </View>
  );
}

test('renders correctly', () => {
  const tree = render(
    <ThemeProvider>
      <ConsumerComponent />
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

test('provides theme correctly', () => {
  const { getByText, getByTestId } = render(
    <ThemeProvider>
      <ConsumerComponent />
    </ThemeProvider>,
  );

  expect(getByTestId('isDarkMode')).toHaveTextContent('Light Mode');
  fireEvent.press(getByText('toggle'));
  expect(getByTestId('isDarkMode')).toHaveTextContent('Dark Mode');
});

describe('with dark mode set as device theme', () => {
  beforeAll(() => {
    jest.spyOn(Appearance, 'getColorScheme').mockReturnValue('dark');
  });
  it('renders with dark mode', () => {
    const tree = render(
      <ThemeProvider>
        <ConsumerComponent />
      </ThemeProvider>,
    );
    expect(tree.getByTestId('isDarkMode')).toHaveTextContent('Dark Mode');
  });
});
