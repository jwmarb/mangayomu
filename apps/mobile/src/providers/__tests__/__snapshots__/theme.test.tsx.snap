// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View>
  <Text
    testID="isDarkMode"
  >
    Light Mode
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        toggle
      </Text>
    </View>
  </View>
</View>
`;
